File: gems/decomplex/rust/src/decomplex/syntax/adapters/swift.rs
use super::super::Language;
use super::base::LanguageProfile;
use crate::decomplex::ast::{node_text, normalize_text};
use tree_sitter::{Language as TreeSitterLanguage, Node};

pub(crate) struct SwiftProfile;

impl LanguageProfile for SwiftProfile {
    fn language(&self) -> Language {
        Language::Swift
    }

    fn grammar(&self) -> TreeSitterLanguage {
        tree_sitter_swift::LANGUAGE.into()
    }

    fn function_node_kinds(&self) -> &[&str] {
        &["function_declaration"]
    }

    fn class_owner_node_kinds(&self) -> &[&str] {
        &["class_declaration"]
    }

    fn parameter_list_node_kinds(&self) -> &[&str] {
        &["function_value_parameters"]
    }

    fn parameter_identifier_node_kinds(&self) -> &[&str] {
        &["simple_identifier"]
    }

    fn inline_parameter_node_kinds(&self) -> &[&str] {
        &["parameter"]
    }

    fn function_body_node_kinds(&self) -> &[&str] {
        &["function_body", "statements"]
    }

    fn nested_statement_wrapper_node_kinds(&self) -> &[&str] {
        &["statements"]
    }

    fn call_node_kinds(&self) -> &[&str] {
        &["call_expression"]
    }

    fn identifier_node_kinds(&self) -> &[&str] {
        &["simple_identifier", "type_identifier"]
    }

    fn assignment_node_kinds(&self) -> &[&str] {
        &["assignment"]
    }

    fn expression_list_node_kinds(&self) -> &[&str] {
        &["directly_assignable_expression"]
    }

    fn assignment_operator_tokens(&self) -> &[&str] {
        &["=", "+=", "-=", "*=", "/=", "%="]
    }

    fn path_action_node_kinds(&self) -> &[&str] {
        &["call_expression", "control_transfer_statement"]
    }

    fn simple_action_wrapper_node_kinds(&self) -> &[&str] {
        &["statements", "control_structure_body", "function_body"]
    }

    fn local_identifier_wrapper_node_kinds(&self) -> &[&str] {
        &[
            "directly_assignable_expression",
            "value_argument",
            "pattern",
        ]
    }

    fn local_declaration_node_kinds(&self) -> &[&str] {
        &["property_declaration", "variable_declaration"]
    }

    fn variable_declaration_node_kinds(&self) -> &[&str] {
        &["variable_declaration", "directly_assignable_expression"]
    }

    fn comparison_node_kinds(&self) -> &[&str] {
        &[
            "equality_expression",
            "comparison_expression",
            "conjunction_expression",
            "additive_expression",
            "multiplicative_expression",
        ]
    }

    fn branch_node_kinds(&self) -> &[&str] {
        &["if_statement", "for_statement", "switch_statement"]
    }

    fn case_node_kinds(&self) -> &[&str] {
        &["switch_statement"]
    }

    fn case_arm_node_kinds(&self) -> &[&str] {
        &["switch_entry"]
    }

    fn case_pattern_node_kinds(&self) -> &[&str] {
        &["switch_pattern", "pattern"]
    }

    fn case_container_stop_node_kinds(&self) -> &[&str] {
        &["function_declaration", "class_declaration"]
    }

    fn case_subject_skip_node_kinds(&self) -> &[&str] {
        &["switch_entry", "else", "comment"]
    }

    fn default_case_patterns(&self) -> &[&str] {
        &["_", "default"]
    }

    fn boolean_and_operators(&self) -> &[&str] {
        &["&&", "and"]
    }

    fn boolean_container_node_kinds(&self) -> &[&str] {
        &[
            "conjunction_expression",
            "equality_expression",
            "comparison_expression",
        ]
    }

    fn boolean_wrapper_node_kinds(&self) -> &[&str] {
        &["statements", "pattern"]
    }

    fn navigation_suffix_node_kinds(&self) -> &[&str] {
        &["navigation_suffix"]
    }

    fn field_like_node_kinds(&self) -> &[&str] {
        &["navigation_expression"]
    }

    fn call_argument_texts(&self, node: Node<'_>, source: &str) -> Vec<String> {
        self.call_argument_nodes(node)
            .into_iter()
            .filter_map(|argument| {
                let text = normalize_text(node_text(argument, source));
                let value = text
                    .strip_prefix('(')
                    .and_then(|inner| inner.strip_suffix(')'))
                    .unwrap_or(&text)
                    .trim()
                    .to_string();
                (!value.is_empty()).then_some(value)
            })
            .collect()
    }
}

File: gems/fact-mine/lib/fact_mine/syntax/swift.rb
# frozen_string_literal: true

module FactMine
  module Syntax
    SWIFT_LEXICON = LanguageLexicon.new(
      nil_literal_patterns: [/\bnil\b/].freeze,
      type_guard_patterns: [
        /\bnil\b/,
        /(?:\?\.|\?\?)/,
        /\b(?:if|guard)\s+let\b/,
        /\b(?:as\?|is)(?:\s|$)/
      ].freeze,
      diagnostic_patterns: [
        /\bthrow\b/,
        /\b(?:fatalError|preconditionFailure|assertionFailure|assert|precondition)\s*\(/
      ].freeze,
      trivial_patterns: [
        /\A(?:nil|true|false|0|1|break|continue)\s*;?\z/,
        /\Areturn\s+(?:nil|true|false|0|1)\s*;?\z/
      ].freeze
    ).freeze

    class SwiftSyntaxAdapter < TreeSitterLanguageAdapter
      FUNCTION_NODE_KINDS = %w[function_declaration].freeze
      CALL_NODE_KINDS = %w[call_expression].freeze
      ADJACENT_CALL_NODE_KINDS = %w[navigation_expression directly_assignable_expression simple_identifier].freeze
      CLASS_OWNER_NODE_KINDS = %w[class_declaration].freeze
      PARAMETER_LIST_NODE_KINDS = %w[function_value_parameters].freeze
      INLINE_PARAMETER_NODE_KINDS = %w[parameter].freeze
      FUNCTION_BODY_NODE_KINDS = %w[function_body statements].freeze
      NESTED_STATEMENT_WRAPPER_NODE_KINDS = %w[statements].freeze
      IDENTIFIER_NODE_KINDS = %w[simple_identifier type_identifier].freeze
      FIELD_IDENTIFIER_NODE_KINDS = [].freeze
      PARAMETER_IDENTIFIER_NODE_KINDS = %w[simple_identifier].freeze
      LOCAL_IDENTIFIER_WRAPPER_NODE_KINDS = %w[directly_assignable_expression value_argument pattern].freeze
      LOCAL_DECLARATION_NODE_KINDS = %w[property_declaration variable_declaration].freeze
      VARIABLE_DECLARATION_NODE_KINDS = %w[variable_declaration directly_assignable_expression].freeze
      LOCAL_VARIABLE_DECLARATOR_NODE_KINDS = [].freeze
      FIELD_DECLARATION_NODE_KINDS = %w[property_declaration].freeze
      DECLARATION_SITE_PARENT_NODE_KINDS = %w[parameter variable_declaration property_declaration function_declaration class_declaration].freeze
      ASSIGNMENT_NODE_KINDS = %w[assignment].freeze
      ASSIGNMENT_STATE_DECLARATION_NODE_KINDS = %w[assignment].freeze
      ASSIGNMENT_OPERATOR_TOKENS = %w[= += -= *= /= %=].freeze
      PATH_ACTION_NODE_KINDS = %w[call_expression control_transfer_statement].freeze
      SIMPLE_ACTION_WRAPPER_NODE_KINDS = %w[statements control_structure_body function_body].freeze
      COMPARISON_NODE_KINDS = %w[equality_expression comparison_expression conjunction_expression additive_expression multiplicative_expression].freeze
      BRANCH_NODE_KINDS = %w[if_statement for_statement switch_statement].freeze
      LOOP_NODE_KINDS = %w[for_statement].freeze
      BRANCH_LOOP_NODE_KINDS = LOOP_NODE_KINDS
      CASE_NODE_KINDS = %w[switch_statement].freeze
      BRANCH_CASE_NODE_KINDS = %w[switch_statement].freeze
      IF_NODE_KINDS = %w[if_statement].freeze
      HIDDEN_IF_WRAPPER_NODE_KINDS = %w[statements].freeze
      HIDDEN_IF_TOKEN_KINDS = %w[if].freeze
      CASE_ARM_NODE_KINDS = %w[switch_entry].freeze
      SWITCH_CASE_ARM_NODE_KINDS = %w[switch_entry].freeze
      CASE_PATTERN_NODE_KINDS = %w[switch_pattern pattern].freeze
      CASE_CONTAINER_STOP_NODE_KINDS = %w[function_declaration class_declaration].freeze
      CASE_SUBJECT_SKIP_NODE_KINDS = %w[switch_entry else comment].freeze
      DEFAULT_CASE_PATTERNS = %w[_ default].freeze
      BOOLEAN_AND_OPERATORS = %w[&& and].freeze
      BOOLEAN_CONTAINER_NODE_KINDS = %w[conjunction_expression equality_expression comparison_expression].freeze
      BOOLEAN_WRAPPER_NODE_KINDS = %w[statements pattern].freeze
      ARGUMENT_LIST_NODE_KINDS = %w[call_suffix value_argument].freeze
      SELF_CALL_IDENTIFIER_NODE_KINDS = %w[simple_identifier type_identifier].freeze
      SELF_RECEIVER_NAMES = %w[self].freeze
      PUBLIC_VISIBILITY_TOKENS = %w[public pub].freeze
      ACCESSOR_CALL_NODE_KINDS = [].freeze
      NAVIGATION_SUFFIX_NODE_KINDS = %w[navigation_suffix].freeze
      FIELD_LIKE_NODE_KINDS = %w[navigation_expression directly_assignable_expression].freeze
      BLOCK_ARGUMENT_NODE_KINDS = [].freeze
    end
  end
end

module FactMine
  module Syntax
    class SwiftNormalizedExtractionBehavior < NormalizedExtractionBehavior
      def self_member_receiver(message)
        "self.#{message}"
      end

      def function_name_from_text(text)
        text.to_s.strip[/\bfunc\s+([A-Za-z_]\w*)\s*\(/, 1] || super
      end

      def access_span_call_site?(message, _current_function)
        message == "fallback"
      end

      def wrap_branch_predicate?(_branch)
        false
      end
    end

    NormalizedExtractionBehavior.register(:swift, SwiftNormalizedExtractionBehavior)
  end
end
